### ISSUE TITLE
mIoU and RI for Multi-/Per- Scan

### ISSUE THREAD

**Issue by jx-zhong-for-academic-purpose** [huangjh-pub/multibody-sync#5]
Many thanks for your wonderful research and this open-source implementation!

Here is a small question -  From [test.py](https://github.com/huangjh-pub/multibody-sync/blob/public/test.py), it seems that I can't find the part of the computation of mIoU and RI for Multi-/Per- Scan. What should I do to obtain the quantitative performance as shown in Table 2 & Table 3?

**Comment by heiwang1997** [comment]
Hi! Thank you so much for your interest in our work!
Here please adapt the following code to get the clustering metric:

First define this class:
```python
from scipy.optimize import linear_sum_assignment

class ClusteringMetrics(nn.Module):

    # Mean IoU       (based on IoU Confusion Matrix)
    IOU = 1

    # Rand Index
    RI = 2

    def __init__(self, spec=None):
        super().__init__()
        if spec is None:
            self.spec = [self.IOU, self.RI]
        else:
            self.spec = spec

    def forward(self, pd_segm: torch.Tensor, segm: torch.Tensor, s: int = None, **kwargs):
        """
        :param segm:    (B, ...) starts from ID 1
        :param pd_segm: (B, ..., s) or (B, ...) starts from ID (0...s-1)
        :param s: the maximum class ID.
        """
        output_dict = {}

        n_batch = pd_segm.size(0)
        gt_segm = segm.reshape(n_batch, -1).detach()                  # (B, N)
        n_data = gt_segm.size(-1)
        n_gt_segms = gt_segm.max(dim=1).values                        # (B,)

        if s is None:
            s = pd_segm.size(-1)
            pd_segm = pd_segm.reshape(n_batch, -1, s).detach()  # (B, N, s)
            pd_segm = pd_segm.argmax(dim=-1)  # (B, N)
        else:
            pd_segm = pd_segm.reshape(n_batch, -1).detach()

        s = max(s, n_gt_segms.max())

        pd_segm = torch.eye(s, dtype=torch.float32,
                            device=pd_segm.device)[pd_segm]  # (B, N, s)
        gt_segm = torch.eye(s, dtype=torch.float32,
                            device=pd_segm.device)[gt_segm - 1]       # (B, N, s)

        matching_score = torch.einsum('bng,bnp->bgp', gt_segm, pd_segm)             # (B, s, s)

        if self.IOU in self.spec:
            union_score = torch.sum(gt_segm, dim=1).unsqueeze(-1) + \
                          torch.sum(pd_segm, dim=1, keepdim=True) - matching_score      # (B, s, s)
            iou_score = matching_score / (union_score + 1e-8)                           # (B, s, s)
            all_mean_ious = []
            for batch_id, n_gt_segm in enumerate(n_gt_segms):
                assert n_gt_segm <= s
                iou_confusion = iou_score[batch_id, :n_gt_segm, :].cpu().numpy()
                row_ind, col_ind = linear_sum_assignment(iou_confusion, maximize=True)
                assert row_ind.size == n_gt_segm
                current_iou = np.mean(iou_confusion[row_ind, col_ind])
                all_mean_ious.append(current_iou)
            output_dict["iou"] = all_mean_ious

        if self.RI in self.spec:
            ri_matrix_gt = torch.bmm(gt_segm, gt_segm.transpose(-1, -2))        # (B, N, N)
            ri_matrix_pd = torch.bmm(pd_segm, pd_segm.transpose(-1, -2))        # (B, N, N)
            ri = torch.sum(ri_matrix_gt == ri_matrix_pd, dim=-1).sum(dim=-1).float() / (n_data * n_data)
            output_dict["ri"] = ri.cpu().numpy().tolist()

        return output_dict
```

Then compute the metric as:
```python
# Build metric
test_metric = ClusteringMetrics(spec=[ClusteringMetrics.IOU, ClusteringMetrics.RI])

# Forward the model
pd_segm, xyz_trans = model(inputs)

# Evaluate.
multi_scan_metric = test_metric(pd_segm, segm)
per_scan_metric = test_metric(pd_segm[0], segm[0])

multi_scan_iou = multi_scan_metric['iou'][0]
multi_scan_ri = multi_scan_metric['ri'][0]

per_scan_iou_avg = np.mean(per_scan_metric['iou'])
per_scan_iou_std = np.std(per_scan_metric['iou'])
per_scan_ri_avg = np.mean(per_scan_metric['ri'])
per_scan_ri_std = np.std(per_scan_metric['ri'])
```

there you go :)

Best regards.

**Comment by jx-zhong-for-academic-purpose** [comment]
Great! Thank you :)

**Comment by Szy-Young** [comment]
Hi @heiwang1997, very thanks for providing the evaluation toolkit! It's very easy to use.
I have a small question:
The provided code computes the metrics on one sample (scene). For the metrics on a whole dataset, shall I just accumulate the `multi_scan_metric, multi_scan_ri, per_scan_iou_avg, per_scan_iou_std, per_scan_ri_avg, per_scan_ri_std` of all samples and average?

**Comment by heiwang1997** [comment]
Yes :)

**Comment by Szy-Young** [comment]
Got it. Thanks~
